Issue 217: Weaver authentication is making unsafe assumptions on the data and failing out incorrectly.

kaladay opened the pull request:
# Description

When an error occurs either a valid error payload is returned or something else is returned. Do not be opinionated about errors payloads here because the fetchUser should not care at this time.

Preserve the original user role and if it is not defined then set the role to ROLE_ANONYMOUS. Only when the payload has a valid Credentials structure should the session role be deleted and the credentials object be modified.

Having a token when the payload is invalid or the role is anonymous  results in an error message getting stuck. Delete the token in the case that the role is anonymous. After all, the anonymous user should not have a token.

Continue to perform the completion steps needed to resolve the fetchUser request regardless of the payload structure.


Fixes #217

## Type of change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Ran using verdaccio via localhost against both good and bad auth URL.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

kaladay committed:
Issue 217: Weaver authentication is making unsafe assumptions on the data and failing out incorrectly.

When an error occurs either a valid error payload is returned or something else is returned.
Do not be opinionated about errors payloads here because the fetchUser should not care at this time.

Preserve the original user role and if it is not defined then set the role to ROLE_ANONYMOUS.
Only when the payload has a valid Credentials structure should the session role be deleted and the credentials object be modified.

Having a token when the payload is invalid or the role is anonymous  results in an error message getting stuck.
Delete the token in the case that the role is anonymous.
After all, the anonymous user should not have a token.

Continue to perform the completion steps needed to resolve the fetchUser request regardless of the payload structure.

kaladay committed:
Issue 217: also use appConfig.anonymousRole rather than hard-coded 'ROLE_ANONYMOUS'.

wwelling commented on app/services/userService.js:
```diff
@@ -10,10 +10,24 @@ core.service("UserService", function ($q, StorageService, User, WsApi) {
-            delete sessionStorage.role;
-            var credentials = angular.fromJson(res.body).payload.Credentials;
-            currentUser.anonymous = credentials.role === appConfig.anonymousRole ? true : false;
+            var body = !res.body ? {} : angular.fromJson(res.body);
```
@kaladay have you considered `!!res.body` and switching the branches? Shouldn't be applicable here but not entirely positive WsApi fetch proxies' proper HTTP response code and response content type.

kaladay replied:
I used `!res.body` to avoid the extra operator in `!!res.body` as a good performance practice. But this is Javascript, does the `!res.body` behave differently than `!!!res.body` aside from having 3 exclamations rather than 1?

kaladay replied:
I changed the logic after fiddling with the annoyance that is JS boolean logic.

wwelling commented on app/services/userService.js:
```diff
@@ -10,10 +10,24 @@ core.service("UserService", function ($q, StorageService, User, WsApi) {
-            var credentials = angular.fromJson(res.body).payload.Credentials;
-            currentUser.anonymous = credentials.role === appConfig.anonymousRole ? true : false;
+            var body = !res.body ? {} : angular.fromJson(res.body);
+            var credentials = { role: !currentUser.role ? appConfig.anonymousRole : currentUser.role };
```
<img width="580" alt="image" src="https://user-images.githubusercontent.com/8352733/205947757-d9f298cd-417d-4392-a0c0-03d6126cce9d.png">

kaladay replied:
I want to believe `!null` and `!undefined` resolve to a non-zero value.
But this is Javascript.
Maybe because it is JS I should switch to the double-negative.

kaladay replied:
That log shows the desired behavior, the `true` is what I want when using `!test` for cases where test is NULL or test is undefined.

kaladay commented on app/services/userService.js:
```diff
@@ -10,10 +10,24 @@ core.service("UserService", function ($q, StorageService, User, WsApi) {
+
+            // Cannot have a token for the anonymous role.
+            if (currentUser.anonymous) {
+              StorageService.delete("token");
```
Should we treat the token as an existing/another bug?
Authentication is failing but a token is being generated.
The user gets stuck with an invalid token in this case.

wwelling replied:
Yeah, it is likely another bug and should be appropriately cross referenced to Weaver-Webservice-Core.

kaladay replied:
Should I remove this added "delete token" logic?

wwelling replied:
Nah, I don't think the UIs are utilizing the token for anonymous users. I am pretty sure that was the choice we made. Can confirm through the auth override to the custom user details in SAGE.

https://github.com/TAMULib/SAGE/blob/main/src/main/java/edu/tamu/sage/auth/service/AppUserCredentialsService.java#L17

Keep the delete session storage token if current user is anonymous even if unnecessary.

kaladay committed:
Issue 217: Swap and adjust the boolean logic to be more JS-safe.

Reverse the bitwise not operations through a double negative.
Add additional checks extra checks.

There is also a ternary being used in the original code that is pointless.
The logic is effectively:
```
  true ?? true : false
```
Get rid of this pointless ternary.

wwelling reviewed (approved):
Code review approved. @rmathew1011 if you could please provide local testing, we can move forward with this update to @wvr/core. We will now need to consider any other changes before rolling out a patch release.

<img width="635" alt="image" src="https://user-images.githubusercontent.com/8352733/205979691-1e9becef-9030-4d41-be05-42ce1b0ee421.png">
